README.md
# homework1

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-12-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-12-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/homework1/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/homework1/compare/0.1.0...0.1.1


test/homework1/core_test.clj
(ns homework1.core-test
  (:require [clojure.test :refer :all]
            [homework1.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/homework1/socket.clj
(ns homework1.socket
  (:require [aleph.tcp :as tcp]
            [gloss.core :as g]
            [gloss.io :as gio]
            [manifold.stream :as s]))

(defn wrap-duplex-stream
  "s is the duplex stream: both client & server"
  [protocol s]
  (let [out (s/stream)] ;; device-sensor.socket.??
    (s/connect
     (s/map #(gio/encode protocol %) out) ;; source
     s)                                   ;; sink
    (s/splice                             ;; put! --> sink --- source -> take!
     out                                  ;; sink
     (gio/decode-stream s protocol))))

(g/defcodec msg-codec [[:byte :byte] :byte])

(defn broadcast [all-streams msg]
  (run! (fn [s] (s/try-put! s msg 1000))
        all-streams))

(def all-server-streams (atom #{}))

(defn close-server [server]
  (.close server))

(defn server []
  (tcp/start-server
   (fn [server-stream _info]
     (println "server get a new client...")
     (swap! all-server-streams conj server-stream)
     (s/consume
      #(do
         (println "收到数据:" %)
         (broadcast @all-server-streams %))
      server-stream))
   {:port 3333}))

(defn client []
  (wrap-duplex-stream
   msg-codec
   @(tcp/client {:host "localhost"
                 :port 3333})))

(comment
  (def s (server))

  (.close s)

  (def c1 (client))

  (s/consume (partial println "c1:") c1)

  (def c2 (client))
  (s/consume (partial println "c2:") c2)

  (s/try-put! c1 [[2 2] 2] 1000)


  )


src/homework1/tutorial.clj
(ns homework1.tutorial
  (:require [cljfx.api :as fx]))

(defmulti event-handler :event/type)



(def length 500)

(def line-count 20)

(def *state (atom (into {}
                        (for [x (range line-count)
                              y (range line-count)]
                          [[x y] 0]))))

(defn move [[x y] color]
  (swap! *state assoc [x y] color))

(defn gen-lines-circles [length line-count margin]
  (let [space (- length (* 2 margin))
        interval (/ space (dec line-count))
        lst    (mapv (fn [idx] (+ margin (* idx interval)))
                     (range line-count))]
    (concat
     (mapv (fn [pos]
             {:fx/type :line
              :start-x margin
              :start-y (int pos)
              :end-x (+ margin space)
              :end-y (int pos)})
           lst)
     (mapv (fn [pos]
             {:fx/type :line
              :start-x (int pos)
              :start-y margin
              :end-x (int pos)
              :end-y (+ margin space)})
           lst)
     (mapv (fn [[x y] [idx-x idx-y]]
             {:fx/type :circle
              :radius 10
              :layout-x x
              :layout-y y
              :opacity (if (zero? (get @*state [idx-x idx-y])) 0.0 1.0)
              :fill (nth ["white" "red" "black"] (get @*state [idx-x idx-y]))
              :on-mouse-clicked (fn [e]
                                  (prn "x:" x "y:" y " get clicked")
                                  (when (zero? (get @*state [idx-x idx-y]))
                                    (swap! *state assoc [idx-x idx-y] 0)))})
           (for [x lst
                 y lst]
             [x y])

           (for [x-idx (range line-count)
                 y-idx (range line-count)]
             [x-idx y-idx])))))


(defmethod event-handler ::touched [e]
  (prn e))

(defmethod event-handler :default [e]
  (prn "unhandled event:" e))


(defn root-view [state]
  {:fx/type :stage
   :height (+ 30 length)
   :width length
   :min-height (+ 30 length)
   :min-width length
   :showing true
   :scene {:fx/type :scene
           :root {:fx/type :pane
                  :children (gen-lines-circles length line-count 10)
                  }}})

(def renderer
  (fx/create-renderer
   :middleware (fx/wrap-map-desc (fn [state]
                                   {:fx/type root-view
                                    :state state}))
   ))

(fx/mount-renderer *state renderer)


src/homework1/ui.clj
(ns homework1.ui
  (:require [cljfx.api :as fx]
            [manifold.stream :as s]
            [homework1.socket :as socket]))

(def length 500)
(def line-count 20)

(def *client (atom nil))
(def *color (atom 1))

(defn init-color [color]
  (reset! *color color))

(def *state (atom (into {}
                        (for [x (range line-count)
                              y (range line-count)]
                          [[x y] 0]))))
(defn move [[[x y] color]]
  (println "move msg received..." [x y] " color:" color)
  (swap! *state assoc [x y] color))

(defn init-socket []
  (println "socket is init.....")
  (let [client (socket/client)]
    (s/consume move client)
    (reset! *client client)))

(defn gen-lines-circles [length line-count margin]
  (let [space (- length (* 2 margin))
        interval (/ space (dec line-count))
        lst    (mapv (fn [idx] (+ margin (* idx interval)))
                     (range line-count))]
    (concat
     (mapv (fn [pos]
             {:fx/type :line
              :start-x margin
              :start-y (int pos)
              :end-x (+ margin space)
              :end-y (int pos)})
           lst)
     (mapv (fn [pos]
             {:fx/type :line
              :start-x (int pos)
              :start-y margin
              :end-x (int pos)
              :end-y (+ margin space)})
           lst)
     (mapv (fn [[x y] [idx-x idx-y]]
             {:fx/type :circle
              :radius 10
              :layout-x x
              :layout-y y
              :opacity (if (zero? (get @*state [idx-x idx-y])) 0.0 1.0)
              :fill (nth ["white" "red" "black"] (get @*state [idx-x idx-y]))
              :on-mouse-clicked (fn [e]
                                  (prn "x:" x "y:" y " get clicked")
                                  (s/try-put! @*client [[idx-x idx-y] @*color ] 1000)
                                  (when (zero? (get @*state [idx-x idx-y]))
                                    (swap! *state assoc [idx-x idx-y] @*color)))})
           (for [x lst
                 y lst]
             [x y])

           (for [x-idx (range line-count)
                 y-idx (range line-count)]
             [x-idx y-idx])))))

(defn root-view [state]
  {:fx/type :stage
   :height (+ 30 length)
   :width length
   :min-height (+ 30 length)
   :min-width length
   :showing true
   :scene {:fx/type :scene
           :root {:fx/type :pane
                  :children (gen-lines-circles length line-count 10)
                  }}})

(defn renderer []
  (fx/create-renderer
   :middleware (fx/wrap-map-desc (fn [state]
                                   {:fx/type root-view
                                    :state state}))
   ))




(defn init-client [color]
  (init-socket)
  (init-color color)
  (fx/mount-renderer *state (renderer))
  )
